src/randomWallpaperMenu.ts

```diff
@@ -261,25 +239,10 @@ class RandomWallpaperMenu {
          * @param {HistoryEntry} entry The selected history entry
          */
         const onSelect = (entry: HistoryEntry): void => {
-            // Make sure no other preview or reset event overwrites our setWallpaper!
-            this._wallpaperController.prohibitNewWallpaper = true;
-
+            Logger.debug(`Selected entry from history as wallpaper (source: ${entry.source.source})`);
             this._wallpaperController.setWallpaper(entry).then(() => {
-                this._wallpaperController.prohibitNewWallpaper = false;
-
-                if (this._settings.getInt('change-type') as Mode === Mode.LOCKSCREEN && this._savedBackgroundUri) {
-                    // Reset background after previewing the lock screen options
-                    this._wallpaperController.resetWallpaper(this._savedBackgroundUri);
-                } else {
-                    // Update saved background with newly set background image
-                    // so we don't revert to an older state when closing the menu
-                    const backgroundSettings = new Settings.Settings('org.gnome.desktop.background');
-                    this._savedBackgroundUri = backgroundSettings.getString('picture-uri');
-                }
-
```

Comment:
Good catch, thanks! 

I simply reset the preview before setting the wallpaper in the fixup commit now. That seems to work for me when artificially prolonging the defaultWallpaperManager. Can you verify?

Note: This isn't really a nice solution and maybe I should take another look at reworking the logic to prevent this kind of race conditions altogether. But for now, I'd be fine with this quick fix.

Reply:
Yes, works.